File: Java/04. Association Relationship/Practice10.java
/**
 * Practice 10: 1-1 Association — Customer & Credit Card
 * Task: Manage customer profiles with associated credit cards.
 *       Card number must be digits only. Expiration date validation.
 *       Credit limit 500K. Track available credit. Age minimum 18.
 *
 * How to compile and run:
 *   javac Practice10.java
 *   java Practice10
 *
 * Key Concepts:
 *   - 1-1 Association (Customer has a CreditCard)
 *   - Input validation (digit-only card number, age check)
 *   - Date-based expiration validation
 */

import java.time.LocalDate;
import java.time.Period;

public class Practice10 {

    /** CreditCard with card number, expiration date, and credit limit. */
    static class CreditCard {
        private static final double MAX_CREDIT_LIMIT = 500000.00;

        private String cardNumber;
        private LocalDate expirationDate;
        private double creditLimit;
        private double totalSpent;

        /** Creates a new CreditCard. Card number must contain only digits. */
        public CreditCard(String cardNumber, LocalDate expirationDate, double creditLimit) {
            if (!cardNumber.matches("\\d+")) {
                throw new IllegalArgumentException("Card number must contain only digits, got: '" + cardNumber + "'");
            }
            if (creditLimit > MAX_CREDIT_LIMIT) {
                throw new IllegalArgumentException("Credit limit cannot exceed " + MAX_CREDIT_LIMIT);
            }
            this.cardNumber = cardNumber;
            this.expirationDate = expirationDate;
            this.creditLimit = creditLimit;
            this.totalSpent = 0;
        }

        public CreditCard(String cardNumber, LocalDate expirationDate) {
            this(cardNumber, expirationDate, MAX_CREDIT_LIMIT);
        }

        public boolean isValid() {
            return !LocalDate.now().isAfter(expirationDate);
        }

        public double availableCredit() {
            return creditLimit - totalSpent;
        }

        public double outstandingBalance() {
            return totalSpent;
        }

        public void makePurchase(double amount) {
            if (amount <= 0) {
                System.out.println("  [Error] Purchase amount must be greater than 0.");
                return;
            }
            if (!isValid()) {
                System.out.println("  [Error] Credit card has expired. Cannot make purchase.");
                return;
            }
            if (amount > availableCredit()) {
                System.out.printf("  [Error] Purchase of %.2f exceeds available credit (%.2f).%n",
                        amount, availableCredit());
                return;
            }
            totalSpent += amount;
            System.out.printf("  [OK] Purchase of %.2f successful.%n", amount);
        }

        public void showInfo() {
            String status = isValid() ? "Valid" : "Expired";
            System.out.println("  Card Number        : " + cardNumber);
            System.out.println("  Expiration Date    : " + expirationDate);
            System.out.println("  Status             : " + status);
            System.out.printf("  Credit Limit       : %.2f%n", creditLimit);
            System.out.printf("  Total Spent        : %.2f%n", totalSpent);
            System.out.printf("  Available Credit   : %.2f%n", availableCredit());
            System.out.printf("  Outstanding Balance: %.2f%n", outstandingBalance());
        }
    }

    /** Customer with personal info and an associated CreditCard (1-1). */
    static class Customer {
        private static final int MIN_AGE = 18;

        private String name;
        private LocalDate dateOfBirth;
        private CreditCard creditCard;

        /** Creates a Customer. Must be at least 18 years old. */
        public Customer(String name, LocalDate dateOfBirth, CreditCard creditCard) {
            int age = calculateAge(dateOfBirth);
            if (age < MIN_AGE) {
                throw new IllegalArgumentException(
                        "Customer must be at least " + MIN_AGE + " years old. " + name + " is " + age + " years old.");
            }
            this.name = name;
            this.dateOfBirth = dateOfBirth;
            this.creditCard = creditCard;
        }

        private static int calculateAge(LocalDate dob) {
            return Period.between(dob, LocalDate.now()).getYears();
        }

        public int getAge() {
            return calculateAge(dateOfBirth);
        }

        public void makePurchase(double amount) {
            System.out.printf("  %s making purchase of %.2f...%n", name, amount);
            creditCard.makePurchase(amount);
        }

        public void showInfo() {
            System.out.println("  Customer Name      : " + name);
            System.out.println("  Date of Birth      : " + dateOfBirth);
            System.out.println("  Age                : " + getAge());
            creditCard.showInfo();
            System.out.println();
        }
    }

    public static void main(String[] args) {
        // --- Create credit cards ---
        CreditCard card1 = new CreditCard("4532123456789012", LocalDate.of(2028, 12, 31));
        CreditCard card2 = new CreditCard("5678901234567890", LocalDate.of(2025, 6, 15)); // Expired

        // --- Create customers ---
        Customer customer1 = new Customer("Tareq", LocalDate.of(1990, 5, 15), card1);
        Customer customer2 = new Customer("Afsana", LocalDate.of(1995, 8, 20), card2);

        System.out.println("=== Customer Info ===");
        customer1.showInfo();
        customer2.showInfo();

        // --- Make purchases ---
        System.out.println("=== Purchases ===");
        customer1.makePurchase(150000);
        customer1.makePurchase(200000);
        customer1.makePurchase(200000); // Should fail
        System.out.println();

        customer2.makePurchase(50000); // Should fail — expired
        System.out.println();

        System.out.println("=== After Purchases ===");
        customer1.showInfo();

        // --- Invalid card number ---
        System.out.println("=== Invalid Card Number ===");
        try {
            new CreditCard("ABCD-1234", LocalDate.of(2028, 1, 1));
        } catch (IllegalArgumentException e) {
            System.out.println("  [Error] " + e.getMessage());
        }
        System.out.println();

        // --- Underage customer ---
        System.out.println("=== Underage Customer ===");
        try {
            CreditCard youngCard = new CreditCard("1111222233334444", LocalDate.of(2030, 1, 1));
            new Customer("Junior", LocalDate.of(2015, 1, 1), youngCard);
        } catch (IllegalArgumentException e) {
            System.out.println("  [Error] " + e.getMessage());
        }
    }
}

File: Golang/04. Association Relationship/practice-10.go
/**
 * Practice 10: 1-1 Association — Customer & Credit Card
 * Task: Manage customer profiles with associated credit cards.
 *       Card number must be digits only. Expiration date validation.
 *       Credit limit 500K. Track available credit. Age minimum 18.
 *
 * How to run:
 *   go run practice-10.go
 *
 * Key Concepts:
 *   - 1-1 Association (Customer has a CreditCard)
 *   - Input validation (digit-only card number, age check)
 *   - Date-based expiration validation
 */

package main

import (
	"fmt"
	"time"
	"unicode"
)

/** CreditCard with card number, expiration date, and credit limit. */
type CreditCard10 struct {
	CardNumber     string
	ExpirationDate time.Time
	CreditLimit    float64
	TotalSpent     float64
}

const maxCreditLimit10 = 500000.00

/** NewCreditCard creates a new CreditCard. Card number must contain only digits. */
func NewCreditCard10(cardNumber string, expirationDate time.Time, creditLimit float64) (*CreditCard10, error) {
	for _, c := range cardNumber {
		if !unicode.IsDigit(c) {
			return nil, fmt.Errorf("card number must contain only digits, got: '%s'", cardNumber)
		}
	}
	if creditLimit > maxCreditLimit10 {
		return nil, fmt.Errorf("credit limit cannot exceed %.2f", maxCreditLimit10)
	}
	return &CreditCard10{
		CardNumber:     cardNumber,
		ExpirationDate: expirationDate,
		CreditLimit:    creditLimit,
	}, nil
}

func (c *CreditCard10) IsValid() bool {
	return !time.Now().After(c.ExpirationDate)
}

func (c *CreditCard10) AvailableCredit() float64 {
	return c.CreditLimit - c.TotalSpent
}

func (c *CreditCard10) OutstandingBalance() float64 {
	return c.TotalSpent
}

func (c *CreditCard10) MakePurchase(amount float64) {
	if amount <= 0 {
		fmt.Println("  [Error] Purchase amount must be greater than 0.")
		return
	}
	if !c.IsValid() {
		fmt.Println("  [Error] Credit card has expired. Cannot make purchase.")
		return
	}
	if amount > c.AvailableCredit() {
		fmt.Printf("  [Error] Purchase of %.2f exceeds available credit (%.2f).\n",
			amount, c.AvailableCredit())
		return
	}
	c.TotalSpent += amount
	fmt.Printf("  [OK] Purchase of %.2f successful.\n", amount)
}

func (c *CreditCard10) ShowInfo() {
	status := "Valid"
	if !c.IsValid() {
		status = "Expired"
	}
	fmt.Println("  Card Number        :", c.CardNumber)
	fmt.Println("  Expiration Date    :", c.ExpirationDate.Format("2006-01-02"))
	fmt.Println("  Status             :", status)
	fmt.Printf("  Credit Limit       : %.2f\n", c.CreditLimit)
	fmt.Printf("  Total Spent        : %.2f\n", c.TotalSpent)
	fmt.Printf("  Available Credit   : %.2f\n", c.AvailableCredit())
	fmt.Printf("  Outstanding Balance: %.2f\n", c.OutstandingBalance())
}

/** Customer with personal info and an associated CreditCard (1-1). */
type Customer10 struct {
	Name        string
	DateOfBirth time.Time
	CreditCard  *CreditCard10
}

const minAge10 = 18

func calculateAge10(dob time.Time) int {
	now := time.Now()
	age := now.Year() - dob.Year()
	if now.YearDay() < dob.YearDay() {
		age--
	}
	return age
}

/** NewCustomer creates a Customer. Must be at least 18 years old. */
func NewCustomer10(name string, dob time.Time, card *CreditCard10) (*Customer10, error) {
	age := calculateAge10(dob)
	if age < minAge10 {
		return nil, fmt.Errorf("customer must be at least %d years old. %s is %d years old",
			minAge10, name, age)
	}
	return &Customer10{Name: name, DateOfBirth: dob, CreditCard: card}, nil
}

func (c *Customer10) GetAge() int {
	return calculateAge10(c.DateOfBirth)
}

func (c *Customer10) MakePurchase(amount float64) {
	fmt.Printf("  %s making purchase of %.2f...\n", c.Name, amount)
	c.CreditCard.MakePurchase(amount)
}

func (c *Customer10) ShowInfo() {
	fmt.Println("  Customer Name      :", c.Name)
	fmt.Println("  Date of Birth      :", c.DateOfBirth.Format("2006-01-02"))
	fmt.Println("  Age                :", c.GetAge())
	c.CreditCard.ShowInfo()
	fmt.Println()
}

func main() {
	// --- Create credit cards ---
	card1, _ := NewCreditCard10("4532123456789012", time.Date(2028, 12, 31, 0, 0, 0, 0, time.Local), maxCreditLimit10)
	card2, _ := NewCreditCard10("5678901234567890", time.Date(2025, 6, 15, 0, 0, 0, 0, time.Local), maxCreditLimit10)

	// --- Create customers ---
	customer1, _ := NewCustomer10("Tareq", time.Date(1990, 5, 15, 0, 0, 0, 0, time.Local), card1)
	customer2, _ := NewCustomer10("Afsana", time.Date(1995, 8, 20, 0, 0, 0, 0, time.Local), card2)

	fmt.Println("=== Customer Info ===")
	customer1.ShowInfo()
	customer2.ShowInfo()

	// --- Make purchases ---
	fmt.Println("=== Purchases ===")
	customer1.MakePurchase(150000)
	customer1.MakePurchase(200000)
	customer1.MakePurchase(200000) // Should fail
	fmt.Println()

	customer2.MakePurchase(50000) // Should fail — expired
	fmt.Println()

	fmt.Println("=== After Purchases ===")
	customer1.ShowInfo()

	// --- Invalid card number ---
	fmt.Println("=== Invalid Card Number ===")
	_, err := NewCreditCard10("ABCD-1234", time.Date(2028, 1, 1, 0, 0, 0, 0, time.Local), maxCreditLimit10)
	if err != nil {
		fmt.Println("  [Error]", err)
	}
	fmt.Println()

	// --- Underage customer ---
	fmt.Println("=== Underage Customer ===")
	youngCard, _ := NewCreditCard10("1111222233334444", time.Date(2030, 1, 1, 0, 0, 0, 0, time.Local), maxCreditLimit10)
	_, err = NewCustomer10("Junior", time.Date(2015, 1, 1, 0, 0, 0, 0, time.Local), youngCard)
	if err != nil {
		fmt.Println("  [Error]", err)
	}
}
